includes/controllers/class-message.php

```diff
@@ -67,6 +74,95 @@ public function __construct( $args = [] ) {
 		parent::__construct( $args );
 	}
 
+	/**
+	 * Fetches messages.
+	 *
+	 * @return WP_Rest_Response
+	 */
+	public function fetch_messages() {
+
+		// Check authentication.
+		if ( ! is_user_logged_in() ) {
+			return hp\rest_error( 401 );
+		}
+
+		// Get thread.
+		$thread_ids = hivepress()->request->get_context( 'message_thread_ids' );
+
+		// Check thread.
+		if ( ! $thread_ids ) {
+			return hp\rest_error( 400 );
+		}
+
+		// Get thread messages.
+		$thread_messages = Models\Message::query()->filter(
+			[
+				'id__in' => $thread_ids,
+			]
+		)->order( [ 'sent_date' => 'desc' ] )
+			->get();
+
+		// Set sender ID.
+		$sender_id = null;
+
+		foreach ( $thread_messages as $thread_message ) {
+			if ( $thread_message->get_sender__id() !== get_current_user_id() ) {
+				$sender_id = $thread_message->get_sender__id();
+				break;
+			}
+		}
+
+		// Check sender ID.
+		if ( ! $sender_id ) {
+			return hp\rest_error( 400 );
+		}
+
+		// Get new messages.
+		$messages = Models\Message::query()->filter(
+			[
+				'read'      => 0,
+				'recipient' => get_current_user_id(),
+				'sender'    => $sender_id,
+			]
+		)->order( 'sent_date' )
+			->get();
+
+		// Set response.
+		$response = [];
+
+		// Set output.
+		$output = '';
```

Comment:
Output should be rendered only if the _render parameter is passed, because this is not the main purpose of REST actions (to render HTML), just a temporary workaround until we move to Vue.js

The action should return something (e.g. messages as arrays in "data") with an extra "html" parameter if _render is sent with the request

Reply:
this is how it already works, if you scroll down - you'll see that there is condition to check if output variable is not empty